lua/bson_lua_wrapper.c

```diff
@@ -0,0 +1,372 @@
+
+        lua_getfield(L, -1, "value"); //Stack: [bson_table, key, {type, value}, value]
+        switch ((element_type)type) {
+            case TYPE_DOCUMENT: {
+                if (!lua_istable(L, -1)) {
+                    snprintf(errorMessage, 255, "Expected table, but value was of type: %s", 
+                             lua_typename(L, lua_type(L, -1)));
+                    return -1;
+                }
+                BsonObject value;
+                bson_object_initialize_default(&value);
+
+                int success = table_to_bson_object(L, &value, errorMessage);
+                if (success != 0) {
+                    return success;
+                }
+
+                bson_object_put_object(obj, key, &value);
+                break;
+            }
+            case TYPE_ARRAY: {
+                if (!lua_istable(L, -1)) {
+                    snprintf(errorMessage, 255, "Expected array, but value was of type: %s", 
+                             lua_typename(L, lua_type(L, -1)));
+                    return -1;
+                }
+                BsonArray value;
+                bson_array_initialize(&value, 4);
+
+                int success = table_to_bson_array(L, &value, errorMessage);
+                if (success != 0) {
+                    return success;
+                }
+
+                bson_object_put_array(obj, key, &value);
+                break;
+            }
+            case TYPE_INT32: {
+                lua_Integer value = luaL_checkinteger(L, -1);
+                bson_object_put_int32(obj, key, (int32_t)value);
+                break;
+            }
+            case TYPE_INT64: {
+                lua_Integer value = luaL_checkinteger(L, -1);
+                bson_object_put_int64(obj, key, (int64_t)value);
+                break;
+            }
+            case TYPE_STRING: {
+                char *value = strdup(luaL_checkstring(L, -1));
+                bson_object_put_string(obj, key, value);
+                free(value);
+                break;
+            }
+            case TYPE_DOUBLE: {
+                lua_Number value = luaL_checknumber(L, -1);
+                bson_object_put_double(obj, key, (double)value);
+                break;
+            }
+            case TYPE_BOOLEAN: {
+                if (!lua_isboolean(L, -1)) {
+                    snprintf(errorMessage, 255, "Expected boolean, but value was of type: %s", 
+                             lua_typename(L, lua_type(L, -1)));
+                    return -1;
+                }
+                int value = lua_toboolean(L, -1);
+                bson_object_put_bool(obj, key, (bson_boolean)value);
+                break;
+            }
+            default: {
+                snprintf(errorMessage, 255, "Error parsing table, unrecognized type: %i", (int)type);
+                return -1;
+            }
+        }
+        lua_pop(L, 2); //Stack: [bson_table, key]
+    }
+    lua_pop(L, 2); //Stack: []
```

Comment:
@jacobkeeler At the moment when this line is executed lua stack is contains only `[bson_table]` because `key` was poped by function `lua_next(L, -2)`  (while loop exit criteria in line 54) so to pop one value from lua stack is enough here.
Besides it function `table_to_bson_object()` is used  recursively and should leave lua stack state as it was before its call, in other way it can lead to errors during recursion (for example: some data on lua stack is needed after recursed call of function, but it was poped from stack during recursed call of function)  

So I would propose remove line `lua_pop(L, 2); //Stack: []`at all.